server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -144,111 +145,114 @@ public RepositoriesService(
     /**
      * Registers new repository in the cluster
      * <p>
-     * This method can be only called on the master node. It tries to create a new repository on the master
-     * and if it was successful it adds new repository to cluster metadata.
+     * This method can be only called on the master node.
+     * It tries to create a new repository on the master, and if it was successful, it adds a new repository to cluster metadata.
      *
      * @param request  register repository request
-     * @param listener register repository listener
+     * @param responseListener register repository listener
      */
-    public void registerRepository(final PutRepositoryRequest request, final ActionListener<AcknowledgedResponse> listener) {
+    public void registerRepository(final PutRepositoryRequest request, final ActionListener<AcknowledgedResponse> responseListener) {
         assert lifecycle.started() : "Trying to register new repository but service is in state [" + lifecycle.state() + "]";
         validateRepositoryName(request.name());
 
         // Trying to create the new repository on master to make sure it works
         try {
             validateRepositoryCanBeCreated(request);
         } catch (Exception e) {
-            listener.onFailure(e);
+            responseListener.onFailure(e);
             return;
         }
 
-        final ListenableFuture<AcknowledgedResponse> acknowledgementStep = new ListenableFuture<>();
-        final ListenableFuture<Boolean> publicationStep = new ListenableFuture<>(); // Boolean==changed.
+        // Aggregated result of two asynchronous operations when the cluster acknowledged and state changed
+        record RegisterRepositoryTaskResult(AcknowledgedResponse ackResponse, boolean changed) {}
+
+        SubscribableListener
 
-        if (request.verify()) {
+            .<Void>newForked(validationStep -> {
+                if (request.verify()) {
+                    validatePutRepositoryRequest(request, validationStep);
```

Comment:
Can we have this done together with `validateRepositoryCanBeCreated`? I think it feels nature that pre-cluster-state-update validation/verification is done in one place. Also seems to be wasteful to create the repository twice. 

While I appreciate that `SubscribableListener` generally makes code easier to read, can we keep it separate from this PR? If we move this step into `validateRepositoryCanBeCreated`, I think we don't need to touch the code here at all in this PR which would make it much easier to review. We can always have a follow-up which will be pure-refactoring for introucing `SubscribableListener`.

Reply:
`validateRepositoryCanBeCreated` method is used in other places, I would rather remove it from here. But it breaks unrelated tests. I can try to chase these down in this PR, or following PR.

About SubscribableListener, I found without little refactor it looks worse. May be diff will be a bit prettier, but final code not. I can do "ugly" change and then refactor in another PR if its a preferable way.

Reply:
> About SubscribableListener, I found without little refactor it looks worse. May be diff will be a bit prettier, but final code not. I can do "ugly" change and then refactor in another PR if its a preferable way.

I think it is better to keep functional change and style change separate unless it is very trivial. If somehow we need to revert the functional change, it is also much cleaner if they are separate. This likely won't happen for this PR. But it feels like a overall good principle. 

> validateRepositoryCanBeCreated method is used in other places

In that case, we don't have to literally move the new code into it. It can sit besides it as a new method. My main point is to _not_ touch the below steps of listeners in this PR.

Let's hear what @DaveCTurner has to say about this before changing anything. Thanks!

Reply:
Btw, the `SubscribableListener` change does _not_ have to come _after_ this PR. It would be total reasonable to have a separate PR just for that and merge it before proceeding here. My main point is to have them separate, order is not important.

Reply:
You are absolutely right about splitting it apart, I talked with David before about it, but got tunnel-visioned chasing down test failures and forgot about it. I rethink what I said before, and I believe all of it can be addressed.

I created PR with non functional change - #108788

I have some thoughts how to organize `validateRepositoryCanBeCreated` better, will update this PR after refactoring merged.

Reply:
Awesome. That was my main gripe :) Thanks for splitting it out. 👍

Reply:
updated PR with refactoring included